docs/source/training.rst

```diff
@@ -0,0 +1,276 @@
+will create a BERT model instance with encoder weights copied from the
+``bert-base-uncased`` model and a randomly initialized sequence
+classification head on top of the encoder with an output size of 2.
+
+This is useful because it allows us to make use of the pre-trained BERT
+encoder and easily train it on whatever sequence classification dataset we
+choose. Models are initialized in ``eval`` mode by default. To train, be sure
+to first call ``model.train()``. When we call the model with the target labels
+as a tensor, the first returned element is the Cross Entropy loss between the
+classifier predictions and the passed labels. The following demonstrates
+how you could instantiate a pre-trained model and perform a single update on
+a dummy sentiment classification batch:
+
+.. code-block:: python
+
+    from transformers import BertTokenizer, BertForSequenceClassification, AdamW
+
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
+    model = BertForSequenceClassification.from_pretrained('bert-base-uncased')
+    model.train()
+    optimizer = AdamW(model.parameters(), lr=5e-5) # any PT optimizer works
+
+    text_batch = ["I love Pixar.", "I don't care for Pixar."]
+    labels = torch.tensor([1,0]).unsqueeze(0)
+    encoding = tokenizer.batch_encode_plus(text_batch, return_tensors='pt',
+                                           pad_to_max_length=True)
+    input_ids = encoding['input_ids']
+    attention_mask = encoding['attention_mask']
+
+    outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
+    loss = outputs[0]
+    loss.backward()
+    optimizer.step()
+
+Of course, you can train on GPU by calling ``to('cuda')`` on the model and
+inputs as usual.
+
+We also provide a few tools to make it easier to apply more complex
+fine-tuning procedures. Here's an example using weight decay and learning
+rate scheduling (omitting the data preparation steps for brevity):
+
+.. code-block:: python
+
+    from transformers import get_linear_schedule_with_warmup
+
+    device = torch.device('cuda')
+    model = BertForSequenceClassification.from_pretrained('bert-base-uncased').to(device)
+    model.train()
+
+    # Prepare optimizer and schedule (linear warmup and decay)
+    no_decay = ['bias', 'LayerNorm.weight']
+    optimizer_grouped_parameters = [
+        {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
+        {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
+    ]
+    optimizer = AdamW(optimizer_grouped_parameters, lr=1e-5)
+    scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps, num_train_steps)
+
+    # iterate over batches of preprocessed training data
+    for batch in train_loader:
+        optimizer.zero_grad()
+        
+        input_ids = batch.input_ids.to(device)
```

Comment:
you can call `batch.to(device)`

Reply:
This is where the "we omit the data preparation part" is annoying. Previously the batch was a dict, now it's something that has an `input_ids` attribute and even a magic method.

If the data is prepared the same way as before, it should call `input_ids = encoding['input_ids']` like before. If it's not, the preparation step should be presented separately (and maybe explain why there is that magic `to` method.